documentation/ArrayExpressionLoweringToFIR.md

```diff
@@ -0,0 +1,315 @@
+
+Short-circuiting requires introducing a CFG of basic blocks which makes
+expression lowering changing the basic block context while emitting FIR.
+The creation of basic block for short-circuiting can comply to the following rules:
+
+While lowering expr `E` in basic block `B1`, expression lowering will
+return with the builder pointing at the end of basic block `Bi` such as:
+- `B1` strictly dominates `Bi` (all paths from `B1` leads to `Bi` and all paths arriving to `Bi` went through `B1`).
+- `Bi` may or may not be `B1`.
+- The SSA value returned by expression lowering is defined in `Bi` and is the value of the expression `E`
+- The CFG graph starting at `B1` and ending in `Bi` is a DAG (considering `fir.loop` is an
+ operation and not a CFG loop).
+
+These rules mainly mean that expression lowering does not have to (and should not) create crazy CFG in order
+to achieve short-circuiting.
+
+These rules should guarantee that we do not have to track the CFG state while lowering an expression other than the current insertion point
+(and maybe some places to load symbols. allocate temps...).
+
+### An array shape does not imply an expression is elemental
+All expressions with an array shape are not elemental, for instance `foo(array)` where `foo` is a user
+function that is not elemental cannot be replaced by call to `foo` on each `array` element.
+
+### Some non-elemental expression can be seen as an elemental expression and an affine map
+As stated in https://github.com/flang-compiler/f18/blob/master/documentation/ArrayComposition.md,
+transformational functions are not elemental but they many of them can simply be viewed as be viewed as
+changing the way their array arguments are to be accessed (i.e/ they are functions operating on the indexes and can be seen in MLIR as affine maps).
+
+### Do not early optimize but do not prevent later optimizations
+Expression lowering from the parse-tree to FIR is not meant to be an optimization pass,
+it should only produce FIR that can be optimized later. For instance, if as stated in ArrayComposition.md,
+it is an end goal to avoid unnecessary temporary arrays, lowering is not the place to perform deep temporary removal optimization.
+It is not entirely clear yet what "can be optimized later" implies.
+The strategy should be to start lowering in a simple way and if it turns out that some optimizations are later not possible, it should be
+preferred to find a way to convey the information that is missing in FIR to enable the optimization rather than to perform the optimization
+while lowering.
+
+### Not all elemental operation should be decomposed
+`merge(x, y, b)` when b is scalar does not require a loop to be implemented.
+It can be implemented as single branch to load the right value or even better, a merge
+on the memory reference.
+
+Question: If lowering were to emit a loop, would later optimization be able to transform this to a simple branch ?
```

Comment:
OK, I think the logic around `MERGE` needs to be more described in my doc because it is the base case of all short-circuit logic (`b.AND.foo()` is equivalent to `MERGE(foo(),.false., b)` and MERGE, as you mentioned fits very nicely to `fir.where`).
So I need to describe the case where we can place the `fir.where` inside the loop and the cases where we need to leave it out for short-circuiting purpose.

Reply:
My interpretation of `MERGE` is that it is a "fancy" LLVM `select`-like instruction, which is like the ternary operation in C.  Perhaps I am missing something though.

Reply:
Do we need a scalar `fir.select`?

Reply:
Wait, I am not sure I follow you question. Isn't `%x` always scalar in `fir.select %x {} otherwise {}` ?
I also thought we could always implement merge as a simple select, but then, Fortran is not C, we do not have to always compute everything.  There are cases we may not want to end-up with an llvm select (typically when the conditions is scalar and the true and false expressions are heavy things).
In `MERGE(foo(), bar(), b)` where everything is scalar, we can skip a function call.
if we lower this to `llvm.select` we are not guaranteed we will short-circuit:
```
%1 = fir.call foo
%2 = fir.call bar
%3 = fir.merge.f32 %1 %2 %b
```
So we can instead lower to:
```
fir.where %b {
  %1 = fir.call foo
  fir.store %1 to %tmp
} otherwise {
  %2 = fir.call bar
  fir.store %2 to %tmp
}
```
When we have a merge with an array condition, we anyway (I think) want to use an `fir.where` inside the `fir.loop` to implement it. For other simpler cases, I agree in the end getting an `llvm.select` would be better than branching, but can't  FIR lowering to LLVM lower fir.where to a `llvm.select` in these cases ?

Reply:
Bad question due to overloaded terminology.  Let's call the hypothetical operation `fir.ternary`.  It would be like:
```
   %1 = fir.ternary %cond, %x, %y : (i1, !T, !T) -> !T
```
So `%1` would be `%x` or `%y` depending on the value of `%cond`.

Reply:
I agree that in the case of short-circuiting, FIR requires a form of control flow (`fir.where` suits) is required.

Reply:
My bad, I read "a scalar `fir.where`" instead of `fir.select`, thanks that for clarifying.
I think we do not need `fir.ternary` now that mlir added [`std.select`](https://github.com/tensorflow/mlir/blob/master/g3doc/Dialects/Standard.md#select-operation) as you mentioned the other time (sorry, I should have mentioned it in the example instead of still using `fir.merge.xx`)
So we can indeed implement `MERGE` with either `fir.where` or this `std.select` and we should be covered for all scenarios.